### PR TITLE
DEV: Move Bookmark modal/component to use d-modal

### PR DESCRIPTION
**Pull request by martin-brennan**
c.f. https://meta.discourse.org/t/converting-modals-from-legacy-controllers-to-new-dmodal-component-api/268057

This also converts the Bookmark component to a Glimmer
component.

**NOTE:** This DE PR needs to be merged after the core one https://github.com/discourse/discourse-data-explorer/pull/252

### CODE REVIEW

**Comment by martin-brennan** [comment]
Plugin frontend failing test will be fixed when the DE PR is merged.

File: app/assets/javascripts/discourse/app/components/bookmark-list.js [hunk 1]
```diff
@@ -84,17 +86,20 @@ export default Component.extend(Scrolling, {
...
-          savedData,
-          bookmark.attachedTo()
-        );
-        this.reload();
-      },
-      onAfterDelete: () => {
-        this.reload();
+    this.modal.show(BookmarkModal, {
+      model: {
+        bookmark: new BookmarkFormData(bookmark),
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/components/bookmark-list.js:91]: This BookmarkFormData thing is not super ideal but I didn't want to go into the rabbit hole of changing Bookmark model to not use RestModel at the same time as well.

**Review by davidtaylorhq — APPROVED** [review]

File: app/assets/javascripts/discourse/app/components/modal/bookmark.hbs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        class="bookmark-name"
+        @enter={{action "saveAndClose"}}
+        placeholder={{i18n "post.bookmarks.name_placeholder"}}
+        maxlength="100"
+        aria-label={{i18n "post.bookmarks.name_input_label"}}
+      />
+      <DButton
+        @icon="cog"
+        @action={{action "toggleShowOptions"}}
+        @class="bookmark-options-button"
```
Comment (janzenisaac) [inline app/assets/javascripts/discourse/app/components/modal/bookmark.hbs:25]:
```suggestion
        @action={{this.toggleShowOptions}}
        class="bookmark-options-button"
```

Comment (janzenisaac) [inline app/assets/javascripts/discourse/app/components/modal/bookmark.hbs:25]:
```suggestion
        @action={{this.toggleShowOptions}}
        class="bookmark-options-button"
```

File: app/assets/javascripts/discourse/app/components/modal/bookmark.hbs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            (i18n "bookmarks.no_timezone" basePath=(base-path))
+          }}</div>
+      {{/if}}
+    </div>
+  </:body>
+
+  <:footer>
+    <div class="control-group">
+      <DButton
+        @id="save-bookmark"
```
Comment (janzenisaac) [inline app/assets/javascripts/discourse/app/components/modal/bookmark.hbs:82]:
```suggestion
        id="save-bookmark"
```

File: app/assets/javascripts/discourse/app/components/modal/bookmark.hbs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        @class="btn-primary"
+        @action={{action "saveAndClose"}}
+      />
+      <DModalCancel @close={{action "closeWithoutSavingBookmark"}} />
+      {{#if this.showDelete}}
+        <DButton
+          @id="delete-bookmark"
+          @icon="trash-alt"
+          @class="delete-bookmark btn-danger"
+          @action={{action "delete"}}
```
Comment (janzenisaac) [inline app/assets/javascripts/discourse/app/components/modal/bookmark.hbs:93]:
```suggestion
          class="delete-bookmark btn-danger"
          @action={{this.delete}}
```

**Review by janzenisaac — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/discourse/app/components/modal/bookmark.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    }
+    return labels;
+  }
+
+  @action
+  unbindKeyboard() {
+    this._itsatrap?.destroy();
+    this._itsatrap = null;
+    KeyboardShortcuts.unpause();
+  }
```
[the remark this answers is not in the document]
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/modal/bookmark.js:137]: Oh I didn't realise that `willDestroy` was left around for the lifecycle, thanks!

**Comment by martin-brennan** [comment]
@janzenisaac I've fixed up all the review comments now, LMK if it's all good :) Thanks for the review!

File: app/assets/javascripts/discourse/app/components/modal/bookmark.js [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+  @service dialog;
+  @service currentUser;
+  @service site;
+
+  @tracked postDetectedLocalDate = null;
+  @tracked postDetectedLocalTime = null;
+  @tracked postDetectedLocalTimezone = null;
+  @tracked prefilledDatetime = null;
+  @tracked flash = null;
+  @tracked flashType = "error";
```
Comment (janzenisaac) [inline app/assets/javascripts/discourse/app/components/modal/bookmark.js:38]: We can drop this here in favor of adding the string directly to the `<DModal>`

```
<DModal @flashType="error">
``` 

as it doesn't look like we are manipulating the value.

**Review by janzenisaac — APPROVED** [review]

**Pull request merged**